### ISSUE TITLE
Re-Read Programming Phoenix Book for v.1.4

### ISSUE THREAD

**Issue by nelsonic** [nelsonic/practice#55]
With the recent release of the Programming Phoenix 1.4 Book (2018-10-19)
see: https://github.com/dwyl/learn-phoenix-framework#book
I feel that it would be a _good_ use of my time to re-read it and follow all the examples
to ensure that my knowledge is up-to-date for my upcoming Phoenix-based web app projects.

If I sit down and focus, I estimate that it will probably take me 2 days 
to read the 269 pages and write the example code.

Better get cracking!

**Comment by nelsonic** [comment]
By the end of page 38 this is what you should see:
![image](https://user-images.githubusercontent.com/194400/51437839-8d70cf80-1c9b-11e9-87d3-72143aa53d5a.png)

**Comment by nelsonic** [comment]
At the end of P. 48 you should see this:
http://localhost:4000/users/
![image](https://user-images.githubusercontent.com/194400/51556122-2d338680-1e71-11e9-817b-7cdee8154c8d.png)

http://localhost:4000/users/2
![image](https://user-images.githubusercontent.com/194400/51556143-3de3fc80-1e71-11e9-9096-c416df75cbdb.png)

**Comment by nelsonic** [comment]
![image](https://user-images.githubusercontent.com/194400/51570898-4ef43400-1e98-11e9-96db-4dacceb2228c.png)

**Comment by nelsonic** [comment]
![image](https://user-images.githubusercontent.com/194400/51571045-e194d300-1e98-11e9-87e9-b3646078cf80.png)

**Comment by nelsonic** [comment]
```
mix phx.routes
Compiling 5 files (.ex)
page_path  GET   /           RumblWeb.PageController :index
user_path  GET   /users      RumblWeb.UserController :index
user_path  GET   /users/new  RumblWeb.UserController :new
user_path  GET   /users/:id  RumblWeb.UserController :show
user_path  POST  /users      RumblWeb.UserController :create
```

**Comment by nelsonic** [comment]
http://localhost:4000/users
![image](https://user-images.githubusercontent.com/194400/51590252-2cd2d400-1ee1-11e9-9c45-9573bee95acf.png)

**Comment by nelsonic** [comment]
p.69 http://localhost:4000/users/new
![image](https://user-images.githubusercontent.com/194400/51592258-0ca61380-1ee7-11e9-8028-c8c4ae2f7830.png)
Form error working. 👍

**Comment by nelsonic** [comment]
Confirm cannot be blank:
![image](https://user-images.githubusercontent.com/194400/51592372-5b53ad80-1ee7-11e9-899c-67504fe5356c.png)

**Comment by nelsonic** [comment]
Seeing the following warning:
![image](https://user-images.githubusercontent.com/194400/51631821-a13f5e80-1f45-11e9-9590-d3f91fae36a5.png)

```
Compiling 2 files (.ex)
warning: invalid association `credential` in schema Rumbl.Accounts.User: associated schema Credential does not exist
  lib/rumbl/accounts/user.ex:1: Rumbl.Accounts.User (module)

warning: function Credential.changeset/2 is undefined (module Credential is not available)
  lib/rumbl/accounts/user.ex:23
```

**Comment by nelsonic** [comment]
Derp ... needed to add the line:
```
```

**Comment by nelsonic** [comment]
Derp ... needed to add the line:
```
alias Rumbl.Accounts.Credential
```
To the top of the `lib/rumbl/accounts/user.ex` file.

No more warnings:
![image](https://user-images.githubusercontent.com/194400/51632026-33476700-1f46-11e9-909e-c4310d008951.png)

**Comment by nelsonic** [comment]
This should work:
```
$ iex -S mix
iex> alias Rumbl.Accounts.User
iex> alias RumblWeb.Router.Helpers, as: Routes
iex> changeset = User.registration_changeset(%User{}, %{ username: "max", name: "Max", credential: %{ password: "asecret", email: "max@example.com"}})
```

![image](https://user-images.githubusercontent.com/194400/51632684-d51b8380-1f47-11e9-8add-2a7ac9a195fe.png)

**Comment by nelsonic** [comment]
```
iex> for u <- Repo.preload(Repo.all(User), :credential) do 
Repo.update!(User.registration_changeset(u, %{ credential: %{email: "#{u.username}@example.com",
password: "temppass"}})) end
```
I ❤️ Elixir!

**Comment by nelsonic** [comment]
http://localhost:4000/users/new
![image](https://user-images.githubusercontent.com/194400/51635433-a654db80-1f4e-11e9-977b-c30d0a2ccccc.png)

Password length error message working.

Works when length meets min required (6):
![image](https://user-images.githubusercontent.com/194400/51635460-ba98d880-1f4e-11e9-8d5c-908b19f8ccdb.png)

HTML is good as Google Chrome asks to save the password. ✅

**Comment by nelsonic** [comment]
Minor detour to read the docs for Plug https://hexdocs.pm/plug/Plug.Conn.html 👀

**Comment by nelsonic** [comment]
on p.88 I'm seeing the following warning:
![image](https://user-images.githubusercontent.com/194400/51638011-2bdb8a00-1f55-11e9-8937-85b5f6813465.png)

**Comment by nelsonic** [comment]
Welcome Back! 🎉 
![image](https://user-images.githubusercontent.com/194400/51640428-5cbebd80-1f5b-11e9-8a4c-dbf922af4b7e.png)

**Comment by nelsonic** [comment]
p.103
```
mix phx.gen.html Multimedia Video videos user_id:references:users \
url:string title:string description:text
```

Add the resource to your browser scope in lib/rumbl_web/router.ex:
```
resources "/videos", VideoController
```

```
$ mix ecto.migrate
```

**Comment by nelsonic** [comment]
![image](https://user-images.githubusercontent.com/194400/51676099-887a8b80-1fcd-11e9-8611-5be234e314a8.png)

This is _expected_ on p.117 ✅

**Comment by nelsonic** [comment]
By p.118 Editing videos works (_as expected_)
![image](https://user-images.githubusercontent.com/194400/51680765-8bc84400-1fda-11e9-968c-b12366ff1f91.png)

**Comment by nelsonic** [comment]
p.123 `mix ecto.gen.migration add_category_id_to_video`
edit the resulting migration and add the alter command:
```
defmodule Rumbl.Repo.Migrations.AddCategoryIdToVideo do
  use Ecto.Migration

  def change do
    alter table(:videos) do
      add :category_id, references(:categories) 
    end
  end
end
```
followed by:
```
mix ecto.migrate
```

**Comment by nelsonic** [comment]
![image](https://user-images.githubusercontent.com/194400/51706284-e0d57b80-2015-11e9-8440-03067e03755f.png)
![image](https://user-images.githubusercontent.com/194400/51706307-f6e33c00-2015-11e9-82cd-993726f37d13.png)

Still does not _display_ category on the `:show` view. `#comingsoon` 👍

**Comment by nelsonic** [comment]
# The Query API (p.131)
Ecto supports a wide range of query options see: https://hexdocs.pm/ecto/Ecto.Query.API.html
+ Comparison operators: ==, !=, <=, >=, <, > • Boolean operators: and, or, not
+ Inclusion operator: in
+ Search functions: like and ilike
+ Null check functions: is_nil
+ Aggregates: count, avg, sum, min, max
+ Date/time intervals: datetime_add, date_add • General: fragment, field, and type
In short, you can use many of the same comparison, inclusion, search, and aggregate operations for a typical query that you’d use in Elixir. You can see documentation and examples for many of them in the Ecto.Query.API documen- tation.2 Those are the basic features you’re going to use as you build queries. You’ll use them from two APIs: keywords syntax and pipe syntax. Let’s see what each API looks like.

**Comment by nelsonic** [comment]
Pipe Syntax queries:
![image](https://user-images.githubusercontent.com/194400/51767896-91a34f80-20d6-11e9-992b-e9e32ffdb172.png)

**Comment by nelsonic** [comment]
The _latest_ version of the Programming Phoenix 1.4 book is considered "content complete":
![image](https://user-images.githubusercontent.com/194400/61172990-a42a1c00-a584-11e9-9120-ca78eabd4f76.png)

So now is a _great_ time to re-read it.

**Comment by nelsonic** [comment]
Reading up on how Ecto Sandbox works: https://hexdocs.pm/ecto_sql/Ecto.Adapters.SQL.Sandbox.html
it's _so_ much better than any other DB adapter I've used. 😮

**Comment by nelsonic** [comment]
# "_If your code is worth writing, it’s worth testing_." 
p.167

**Comment by nelsonic** [comment]
Each time I sit down to try and read (_and follow/**type** the examples_) there are at least 10 GitHub notifications competing for my attention ... 
I _really_ need a "Reading Tracker" https://github.com/nelsonic/time-mvp-phoenix/issues/3 
or more broadly a _general_ purpose "Learning Tracker" [collaborative-learning-community.md](https://github.com/dwyl/product-roadmap/blob/master/collaborative-learning-community.md) ...

My plan is to use Phoenix for building the tools. 
So _finishing_ reading the book is the "egg" 🥚that will hatch into the chicken ... 🐣|> 🐔|> 🚀 

_Must_ stay focussed! 👨‍💻 
20 pages today! 🎯

**Comment by nelsonic** [comment]
http://localhost:4000/manage/videos
![image](https://user-images.githubusercontent.com/194400/62600712-74520800-b8e7-11e9-87bb-393d01274d9e.png)
